lib/service/Service.ts

```diff
@@ -610,19 +610,46 @@ class Service {
    * If price is zero or unspecified a market order will get added.
    */
   public placeOrder = async (
-    args: { pairId: string, price: number, quantity: number, orderId: string, side: number,
-      replaceOrderId: string, immediateOrCancel: boolean },
+    args: { pairId: string, price: number, quantity?: number, orderId: string, side: number,
+      replaceOrderId: string, immediateOrCancel: boolean, all?: boolean },
     callback?: (e: ServicePlaceOrderEvent) => void,
   ) => {
     argChecks.PRICE_NON_NEGATIVE(args);
     argChecks.PRICE_MAX_DECIMAL_PLACES(args);
     argChecks.HAS_PAIR_ID(args);
-    const { pairId, price, quantity, orderId, side, replaceOrderId, immediateOrCancel } = args;
+    const { pairId, price, quantity, orderId, side, replaceOrderId, immediateOrCancel, all } = args;
+
+    let calculatedQuantity: number;
+
+    if (all) {
```

Comment:
Could we extract everything in this if/else conditional and extract it to a separate function that has unit tests?

Reply:
> I'm not able to issue a market buy all order when the peer order's offered quantity is higher than we can fill:
> 
> ```
> ➜  xud git:(feature/buy-sell-all-grpc) ✗ ./bin/xucli orderbook BTC/DAI
> 
> Trading pair: BTC/DAI
> ┌───────────────────────────────────────┬───────────────────────────────────────┐
> │ Buy                                   │ Sell                                  │
> ├───────────────────┬───────────────────┼───────────────────┬───────────────────┤
> │ Quantity          │ Price             │ Price             │ Quantity          │
> ├───────────────────┼───────────────────┼───────────────────┼───────────────────┤
> │                   │                   │ 100000            │ 0.02              │
> └───────────────────┴───────────────────┴───────────────────┴───────────────────┘
> ➜  xud git:(feature/buy-sell-all-grpc) ✗ ./bin/xucli getbalance
> 
> Balance:
> ┌──────────┬───────────────┬────────────────────────────┬───────────────────────────────┐
> │ Currency │ Total Balance │ Channel Balance (Tradable) │ Wallet Balance (Not Tradable) │
> ├──────────┼───────────────┼────────────────────────────┼───────────────────────────────┤
> │ BTC      │ 465.55099411  │ 0.08608198                 │ 465.46491213                  │
> ├──────────┼───────────────┼────────────────────────────┼───────────────────────────────┤
> │ DAI      │ 6000          │ 1500                       │ 4500                          │
> ├──────────┼───────────────┼────────────────────────────┼───────────────────────────────┤
> │ ETH      │ 34.99859449   │ 10                         │ 24.99859449                   │
> ├──────────┼───────────────┼────────────────────────────┼───────────────────────────────┤
> │ USDT     │ 2000          │ 0                          │ 2000                          │
> └──────────┴───────────────┴────────────────────────────┴───────────────────────────────┘
> ➜  xud git:(feature/buy-sell-all-grpc) ✗ ./bin/xucli buy all BTC/DAI mkt
> 9 FAILED_PRECONDITION: DAI outbound balance of 150000000000 is not sufficient for order amount of 200000000000
> ```

How should be behave for such cases?

Reply:
Yes of course but that means I need to make this function public, I'm for now committing it as private, but let me know if making it public to write tests.

Reply:
You can access private methods of a TypeScript class with `class['methodName']`. This shouldn't be used in the actual code but it's good enough for writing tests.

Reply:
Oh, thanks for letting me know @michael1011 that showed how noob I'm with typescript 🤦